### PR TITLE
crypto: add a new pbkdf2 implementation

### PR DESCRIPTION
**Pull request by loubaihui**
We add a new pbkdf2 implementation.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
How about just:
``` diff
--- a/core/tee/sub.mk
+++ b/core/tee/sub.mk
@@ -21,7 +21,9 @@ srcs-y += entry_std.c
 srcs-y += tee_cryp_utl.c
 srcs-$(CFG_CRYPTO_HKDF) += tee_cryp_hkdf.c
 srcs-$(CFG_CRYPTO_CONCAT_KDF) += tee_cryp_concat_kdf.c
+ifneq($(CFG_CRYPTO_HW_PBKDF2),y)
 srcs-$(CFG_CRYPTO_PBKDF2) += tee_cryp_pbkdf2.c
+endif
 
 ifeq ($(CFG_WITH_USER_TA),y)
 srcs-y += tee_obj.c
```
instead?

We should also document the `CFG_CRYPTO_HW_PBKDF2` flag, probably in `mk/config.mk`

**Comment by loubaihui** [comment]
> How about just:
> 
> ```diff
> --- a/core/tee/sub.mk
> +++ b/core/tee/sub.mk
> @@ -21,7 +21,9 @@ srcs-y += entry_std.c
>  srcs-y += tee_cryp_utl.c
>  srcs-$(CFG_CRYPTO_HKDF) += tee_cryp_hkdf.c
>  srcs-$(CFG_CRYPTO_CONCAT_KDF) += tee_cryp_concat_kdf.c
> +ifneq($(CFG_CRYPTO_HW_PBKDF2),y)
>  srcs-$(CFG_CRYPTO_PBKDF2) += tee_cryp_pbkdf2.c
> +endif
>  
>  ifeq ($(CFG_WITH_USER_TA),y)
>  srcs-y += tee_obj.c
> ```
> 
> instead?
> 
> We should also document the `CFG_CRYPTO_HW_PBKDF2` flag, probably in `mk/config.mk`

Sounds great, according to your suggestions, we made the revision.

File: core/tee/tee_cryp_pbkdf2.c [hunk 1]
```diff
@@ -112,3 +112,4 @@ TEE_Result tee_cryp_pbkdf2(uint32_t hash_id, const uint8_t *password,
 	crypto_mac_free_ctx(hmac_parms.ctx);
 	return res;
 }
+
```
Comment (jenswi-linaro) [inline core/tee/tee_cryp_pbkdf2.c:115]: Unrelated change.

File: core/crypto/sub.mk [hunk 2]
```diff
@@ -42,4 +42,5 @@ srcs-$(CFG_CRYPTO_ECB) += sm4-ecb.c
 srcs-$(CFG_CRYPTO_CBC) += sm4-cbc.c
 srcs-$(CFG_CRYPTO_CTR) += sm4-ctr.c
 srcs-$(CFG_CRYPTO_XTS) += sm4-xts.c
+srcs-$(CFG_CRYPTO_HW_PBKDF2) += pbkdf2_hw.c
```
Comment (jenswi-linaro) [inline core/crypto/sub.mk:45]: I don't think `pbkdf2_hw.c` or `pbkdf2_support.h` is needed at all. The hardware implementation can just implement the `tee_cryp_pbkdf2()`

Comment (loubaihui) [reply core/crypto/sub.mk:45]: Is that mean:
If the two files are not needed, the hardware implementation also put in `tee_cryp_pbkdf2.c` and use CFG_CRYPTO_HW_PBKDF2 to distinguish?

Comment (jenswi-linaro) [reply core/crypto/sub.mk:45]: No, a hardware non-stubbed implementation is provided in some other file.
It doesn't make sense to use a stubbed hardware implementation when a working software version is available.

Comment (loubaihui) [reply core/crypto/sub.mk:45]: The hardware implementation will be different from the current software, we will use different hash algorithm etc.
So, I think the  `pbkdf2_hw.c` and `pbkdf2_support.h` is necessary, then specific implementation of `hw_pbkdf2()` function will be added in `core/ drivers/crypto` @jenswi-linaro

Comment (jenswi-linaro) [reply core/crypto/sub.mk:45]: Do you mean that the same input will yield different results? I don't see how a stubbed implementation helps against that.

Comment (loubaihui) [reply core/crypto/sub.mk]: > Do you mean that the same input will yield different results? I don't see how a stubbed implementation helps against that.

I'm sorry for misunderstanding your previous comment. Now, I got it and the PR is ready for another round of review.

File: mk/config.mk [hunk 3]
```diff
@@ -1038,3 +1038,7 @@ CFG_TA_OPTEE_CORE_API_COMPAT_1_1 ?= n
 #              - PerInstance/AttestationTest#
 # Note that this violates GP requirements of HMAC size range.
 CFG_HMAC_64_1024_RANGE ?= n
+
+#Enable a hardware pbkdf2 function
```
Comment (jenswi-linaro) [inline mk/config.mk:1042]: Please add a space after `#`, the same on the line below.

File: mk/config.mk [hunk 4]
```diff
@@ -1038,3 +1038,7 @@ CFG_TA_OPTEE_CORE_API_COMPAT_1_1 ?= n
...
+#By default use standard pbkdf2 implementation
+CFG_CRYPTO_HW_PBKDF2 ?= n
```
[the remark this answers is not in the document]
Comment (jforissier) [reply mk/config.mk:1044]: Yes, I agree

File: mk/config.mk (hunk 3 above)
Comment (loubaihui) [reply mk/config.mk]: Modifications are made.

File: mk/config.mk [hunk 5]
```diff
@@ -1038,3 +1038,7 @@ CFG_TA_OPTEE_CORE_API_COMPAT_1_1 ?= n
 #              - PerInstance/AttestationTest#
 # Note that this violates GP requirements of HMAC size range.
 CFG_HMAC_64_1024_RANGE ?= n
+
+# Enable a hardware pbkdf2 function
+CFG_CRYPTO_HW_PBKDF2 ?= y
+$(eval $(call cfg-depends-all,CFG_CRYPTO_HW_PBKDF2,CFG_CRYPTO_PBKDF2))
```
Comment (jenswi-linaro) [inline mk/config.mk:1044]: Please add a newline character at the end of the file to avoid this ugly warning from git.

File: mk/config.mk [hunk 6]
```diff
@@ -1038,3 +1038,7 @@ CFG_TA_OPTEE_CORE_API_COMPAT_1_1 ?= n
 #              - PerInstance/AttestationTest#
 # Note that this violates GP requirements of HMAC size range.
 CFG_HMAC_64_1024_RANGE ?= n
+
+# Enable a hardware pbkdf2 function
+CFG_CRYPTO_HW_PBKDF2 ?= y
```
Comment (jenswi-linaro) [inline mk/config.mk:1043]: The default value should be `n`.

Comment (loubaihui) [reply mk/config.mk]: OK

**Review by jenswi-linaro — COMMENTED** [review]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by loubaihui** [comment]
@jenswi-linaro @jforissier

**Comment by loubaihui** [comment]
> @loubaihui the code looks good to me so please add:
> 
> `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
> 
> However I would like the commit description changed because it does not add a new implementation, just a config flag to support a hardware implementation. How about:
> 
> ```
> crypto: introduce CFG_CRYPTO_HW_PBKDF2
> 
> Add a new configuration flag to support hardware implementation of
> PBKDF2.
> ```

Ok, thanks! @jforissier

**Comment by jforissier** [comment]
The CI error with "QEMUv8, Xen" is unrelated to this PR, and probably caused by insufficient disk space. This will be addressed separately. Therefore I am merging this, thanks @loubaihui.
